api/src/main/java/org/pmiops/workbench/api/ProfileController.java

```diff
@@ -502,6 +517,44 @@ private boolean userNeverLoggedIn(com.google.api.services.admin.directory.model.
     return ResponseEntity.status(HttpStatus.NO_CONTENT).build();
   }
 
+  @Override
+  public ResponseEntity<Void> updatePageVisits(PageVisit newPageVisit) {
+    User user = userProvider.get();
+    boolean shouldAdd = true;
+    List<org.pmiops.workbench.db.model.PageVisit> oldPageVisits =
+      new ArrayList<>(user.getPageVisits());
+    List<org.pmiops.workbench.db.model.PageVisit> newPageVisits = new ArrayList<>();
+    Timestamp timestamp = new Timestamp(clock.instant().toEpochMilli());
+    for (org.pmiops.workbench.db.model.PageVisit pageVisit : oldPageVisits) {
+      if (pageVisit.getPageId().equals(newPageVisit.getPage())) {
+        pageVisit.setLastVisit(timestamp);
+        shouldAdd = false;
+      }
+      newPageVisits.add(pageVisit);
+    }
+    if (shouldAdd) {
+      org.pmiops.workbench.db.model.PageVisit firstPageVisit =
+        new org.pmiops.workbench.db.model.PageVisit();
+      firstPageVisit.setPageId(newPageVisit.getPage());
+      firstPageVisit.setUser(user);
+      firstPageVisit.setFirstVisit(timestamp);
+      firstPageVisit.setLastVisit(timestamp);
+      newPageVisits.add(firstPageVisit);
+    }
+    user.setPageVisits(new HashSet<>(newPageVisits));
```

Comment:
This logic can be simplified. For example:
```
boolean exists = oldPageVisits.stream().anyMatch(v-> v.getPageId().equals(newPageVisit.getPage()));
```

Or, you could get the old page visit directly and update if it is there:
```
    Optional<org.pmiops.workbench.db.model.PageVisit> oldVisit = oldPageVisits
        .stream()
        .filter(v -> v.getPageId().equals(newPageVisit.getPage()))
        .findFirst();
```
With the optional, that gives you enough information to either update what's there or add a new one.